### PR TITLE
Add a small, customizable sleep to the polling loop

### PR DESCRIPTION
**Pull request by jstnlef**
Addresses https://github.com/amethyst/laminar/issues/186. Basically, we were just spinning excessively where it wasn't needed.

### CODE REVIEW

**Comment by jstnlef** [comment]
Seems like the coverage tool is totally busted.

**Comment by fhaynes** [comment]
Out of curiosity, why sleep and not yield?

**Comment by jstnlef** [comment]
A yield will merely cede control back to the CPU scheduler, which promptly re-enters the hot loop. So it goes from 100% to ~95%. A 1 ms sleep brings it down to a much more reasonable ~4%

**Comment by fhaynes** [comment]
Well, it only promptly re-enters it because it has nothing else to do on the system. I would be curious to see how sleep performs on a heavily-loaded server running VMs. My guess is sleep will gradually take longer and longer than 1ms.

It would be nice to make this configurable with the sleep as the default, so that if people want to ensure performance and tight integration with the OS scheduler, they can.

**Comment by jstnlef** [comment]
I hear what you're saying, but the thing is, we really don't need to be pegging the CPU. At no point do we need to be polling so often that the CPU is spinning.

**Comment by fhaynes** [comment]
Pulling the packets in a way dependent on the time-keeping resolution of the OS is very different than pulling them based on the OS scheduler. As the system becomes busier, the CPU % occupied by the polling loop will decrease to smoothly accommodate other system processes.

When you make it determined by the sleep call, the OS has to do additional bookkeeping, and if it is heavily loaded, it will usually end up sleeping for longer than 1ms. Because the OS cannot even schedule it until it has decided to wake it.
What you meant to be a 1ms sleep may end up being an 8ms sleep one iteration, then 15, then back to 1.
When your OS scheduler is actually in a guest VM process that is itself being scheduled by the hypervisor, it is even more complicated, because your timer resolution can be drastically affected by VMs that have nothing to do with yours. This is a not-uncommon problem with DB clusters and other systems that require high-resolution time-keeping in AWS, etc.

**Comment by jstnlef** [comment]
Made it optional 👍

File: src/net/socket.rs [hunk 1]
```diff
@@ -91,11 +92,21 @@ impl Socket {
     }
 
     /// Entry point to the run loop. This should run in a spawned thread since calls to `poll.poll`
-    /// are blocking.
+    /// are blocking. We will default this to sleeping for 1ms.
     pub fn start_polling(&mut self) {
+        self.start_polling_with_duration(Some(Duration::from_millis(1)))
```
Comment (TimonPost) [inline src/net/socket.rs]: hardcoded value, alert, we might want to move that into the config. I know the user is able to call  `start_polling_with_duration`,  so I don't have a problem with it.

Comment (jstnlef) [reply src/net/socket.rs]: I really wish I could do this in rust... (sigh)
```
pub fn start_polling(&mut self, poll_duration: Option<Duration> = Some(Duration::from_millis(1))) {
```

**Pull request merged**